### PR TITLE
Implement diagnostic for String conversion

### LINKED ISSUE #89856: Recommend `Option::as_deref` when given `Option<&String>` in place of `Option<&str>`
`&String` can easily be converted to `&str`, but we give a somewhat opaque E0308 when faced with the following situation:

```rust
fn take_str_maybe(x: Option<&str>) -> Option<&str> { None }

fn main() {
    let string = String::from("Hello, world");
    let option = Some(&string);
    take_str_maybe(option);
}
```

This results in:

```text
error[E0308]: mismatched types
 --> src/main.rs:8:20
  |
8 |     take_str_maybe(option);
  |                    ^^^^^^ expected `str`, found struct `String`
  |
  = note: expected enum `Option<&str>`
             found enum `Option<&String>`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` due to previous error
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8673cd490aa386a5572a32d04a7eae27)

Which is not totally clear to people who do not understand the `Deref` trait and/or do not know that it is implemented for these types.

We should suggest that the user use `Option::as_deref` to convert between these types.

### PR DESCRIPTION
**Pull request by terrarier2111**
This is my first real contribution to rustc, any feedback is highly appreciated.
This should fix https://github.com/rust-lang/rust/issues/89856

Thanks to @estebank for guiding me.

### CODE REVIEW

**Comment by joshtriplett** [comment]
@terrarier2111 This seems like a great suggestion! Could you please add a test case that triggers the new diagnostic and tests its output?

(That will also make it easier to review.)

**Comment by terrarier2111** [comment]
> @terrarier2111 This seems like a great suggestion! Could you please add a test case that triggers the new diagnostic and tests its output?
> 
> (That will also make it easier to review.)

Is this okay, how i did it?

**Comment by terrarier2111** [comment]
> The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4130922794?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/aa8cf62207bb0a7678150a01ce50dc2dcdbe98ee/checks/4130922794/logs)
> Click to see the possible cause of the failure (guessed by this bot)
> 
> ```
> diff of stderr:
> 
> 1 error[E0308]: mismatched types
> -  --> $DIR/issue89856.rs:6:20
> -   |
> - 6 |     take_str_maybe(option);
> -   |                    ^^^^^^
> -   |                    |
> -   |                    expected `str`, found struct `String`
> -   |                    help: try converting the passed type into a &str: `.map(|x| &**x)`
> -   |
> Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
> -   = note: expected enum `Option<&str>`
> -              found enum `Option<&String>`
> +    |
> +    |
> + LL |     take_str_maybe(option);
> +    |                    |
> +    |                    expected `str`, found struct `String`
> +    |                    expected `str`, found struct `String`
> +    |                    help: try converting the passed type into a &str: `.map(|x| &**x)`
> +    = note: expected enum `Option<&str>`
> +               found enum `Option<&String>`
> 12 
> 13 error: aborting due to previous error
> ---
> To only update this specific test, also pass `--test-args typeck/issue89856.rs`
> 
> error: 1 errors occurred comparing output.
> status: exit status: 1
> command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/typeck/issue89856.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/typeck/issue89856" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/typeck/issue89856/auxiliary"
> ------------------------------------------
> 
> ------------------------------------------
> stderr:
> stderr:
> ------------------------------------------
> error[E0308]: mismatched types
>   --> /checkout/src/test/ui/typeck/issue89856.rs:6:20
>    |
> LL |     take_str_maybe(option);
>    |                    ^^^^^^
>    |                    |
>    |                    expected `str`, found struct `String`
>    |                    help: try converting the passed type into a &str: `.map(|x| &**x)`
>    = note: expected enum `Option<&str>`
>               found enum `Option<&String>`
> 
> error: aborting due to previous error
> ---
> test result: FAILED. 12276 passed; 1 failed; 110 ignored; 0 measured; 0 filtered out; finished in 138.90s
> 
> 
> 
> command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"
> 
> 
> Build completed unsuccessfully in 0:13:49
> ```

What did i do wrong here? i don't understand this.

**Comment by estebank** [comment]
@terrarier2111, to create/update the `.stderr` file you should use `./x.py test src/test/ui --bless`. The `.stderr` files have some conversions that are hard to get right by hand. If you look at the error, it is giving you a diff between what the file was and what it expected. It is pointing at the whole file because of the leading whitespace (because the output in your terminal was `6`, while in the test suite it always gets replaced with `LL`, which is wider, to avoid nasty reformats when changing the tests). I would also rename `issue89856.rs` to `issue-89856.rs`.

File: compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs [hunk 1]
```diff
@@ -232,48 +232,73 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                    if subst.len() != 0 {
+                        if let rustc_middle::ty::subst::GenericArgKind::Type(ty) = subst[0].unpack()
+                        {
+                            let peeled = ty.peel_refs().to_string();
+                            if peeled == "String" {
+                                let ref_cnt = ty.to_string().len() - peeled.len();
+                                let result = format!(".map(|x| &*{}x)", "*".repeat(ref_cnt));
+                                err.span_suggestion(
+                                    expr.span,
+                                    "try converting the passed type into a &str",
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs:295]:
```suggestion
                                    "try converting the passed type into a `&str`",
```

File: compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs [hunk 2]
```diff
@@ -232,48 +232,73 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                if let ty::Adt(_def, subst) = found.kind() {
+                    if subst.len() != 0 {
+                        if let rustc_middle::ty::subst::GenericArgKind::Type(ty) = subst[0].unpack()
+                        {
+                            let peeled = ty.peel_refs().to_string();
+                            if peeled == "String" {
+                                let ref_cnt = ty.to_string().len() - peeled.len();
+                                let result = format!(".map(|x| &*{}x)", "*".repeat(ref_cnt));
+                                err.span_suggestion(
+                                    expr.span,
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs:294]:
```suggestion
                                    expr.span.shrink_to_hi(),
```

File: compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs [hunk 3]
```diff
@@ -232,48 +232,73 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+            {
+                if let ty::Adt(_def, subst) = found.kind() {
+                    if subst.len() != 0 {
+                        if let rustc_middle::ty::subst::GenericArgKind::Type(ty) = subst[0].unpack()
+                        {
+                            let peeled = ty.peel_refs().to_string();
+                            if peeled == "String" {
+                                let ref_cnt = ty.to_string().len() - peeled.len();
+                                let result = format!(".map(|x| &*{}x)", "*".repeat(ref_cnt));
+                                err.span_suggestion(
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs:293]:
```suggestion
                                err.span_suggestion_verbose(
```

File: src/test/ui/typeck/issue-89856.rs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+fn take_str_maybe(x: Option<&str>) -> Option<&str> { None }
+
+fn main() {
+    let string = String::from("Hello, world");
+    let option = Some(&string);
+    take_str_maybe(option);
+                   //~^ ERROR: mismatched types [E0308]
```
Comment (estebank) [inline src/test/ui/typeck/issue-89856.rs:7]:
```suggestion
    //~^ ERROR: mismatched types [E0308]
```

File: compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs [hunk 5]
```diff
@@ -232,48 +232,73 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                            Applicability::MaybeIncorrect,
+                        );
+                    }
+                }
+            } else if found.to_string().starts_with("Option<")
+                && expected.to_string() == "Option<&str>"
+            {
+                if let ty::Adt(_def, subst) = found.kind() {
+                    if subst.len() != 0 {
+                        if let rustc_middle::ty::subst::GenericArgKind::Type(ty) = subst[0].unpack()
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs:287]: Can you maybe `use rustc_middle::ty::subst::GenericArgKind`? We generally import the names we use. I don't think we ever use full paths unless they are very short.

I think there are ways you could shorten these. I *think* that `subst` is a `Vec`. If so, then you can write something like the following and get rid of two levels of indentation:

```rust
if let ty::Adt(_def, [subst]) = found.kind() {
    if let GenericArgKind::Type(ty) = subst.unpack() {
```

**Review by estebank — COMMENTED** [review]
Thank you for doing this!

I left a couple of nitpicks. Could you also squash your commits into one before we merge?

File: compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs (hunk 5 above)
Comment (terrarier2111) [reply compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs:287]: This doesn't seem to work.

**Comment by terrarier2111** [comment]
I am not managing to squash this, i searched in various forums, for example stack overflow, read this wiki entry on squashing: https://rustc-dev-guide.rust-lang.org/git.html#advanced-rebasing
but locally every time i try to rollback some unwanted changes, they just come back and even using git stash only works for a couple of seconds and then i have to redo it again. Idk what I did, exactly i tried various things to squash this, but i really don't understand what i need to do to get it to work. (i tried git rebase -i already)

**Comment by terrarier2111** [comment]
@estebank I squashed all commits, is there anything left to do for me, or is this good to go?

**Comment by estebank** [comment]
Lets land this as is (`str`/`String` conversions are very common), but ideally we'll come back to this and make it work for any two `A` and `B` that `impl AsRef<A> for B`.

@bors r+

**Pull request merged**